tokio/src/runtime/local.rs

```diff
@@ -0,0 +1,527 @@
+//! Runs `!Send` futures on the current thread.
+use crate::runtime::task::{self, JoinHandle, Schedule, UnsendTask, Unsendable};
+
+use std::cell::{Cell, UnsafeCell};
+use std::collections::VecDeque;
+use std::fmt;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::rc::Rc;
+use std::task::{Context, Poll};
+
+use pin_project::pin_project;
+
+/// A group of tasks which are executed on the same thread.
+///
+/// These tasks need not implement `Send`; a local task group provides the
+/// capacity to execute `!Send` futures.
+#[derive(Debug)]
+pub struct TaskGroup {
+    scheduler: Rc<Scheduler>,
+    _not_send_or_sync: PhantomData<*const ()>,
+}
+
+struct Scheduler {
+    /// List of all active tasks spawned onto this executor.
+    ///
+    /// # Safety
+    ///
+    /// Must only be accessed from the primary thread
+    tasks: UnsafeCell<task::OwnedList<Scheduler, Unsendable>>,
+
+    /// Local run queue.
+    ///
+    /// Tasks notified from the current thread are pushed into this queue.
+    ///
+    /// # Safety
+    ///
+    /// References should not be handed out. Only call `push` / `pop` functions.
+    /// Only call from the owning thread.
+    queue: UnsafeCell<VecDeque<UnsendTask<Scheduler>>>,
+}
+
+#[pin_project]
+struct LocalFuture<F> {
+    scheduler: Rc<Scheduler>,
+    #[pin]
+    future: F,
+}
+
+thread_local! {
+    static CURRENT_TASK_SET: Cell<Option<NonNull<Scheduler>>> = Cell::new(None);
+}
+
+/// Spawns a `!Send` future on the local task group.
+///
+/// The spawned future will be run on the same thread that called `spawn.`
+/// This may only be called from the context of a local task group.
+///
+/// # Panics
+///
+/// - This function panics if called outside of a local task group.
+///
+/// # Examples
+///
+/// ```rust
+/// # use tokio::runtime::Runtime;
+/// use std::rc::Rc;
+/// use tokio::local;
+///
+/// let unsend_data = Rc::new("my unsend data...");
+///
+/// let mut rt = Runtime::new().unwrap();
+/// let local_group = local::TaskGroup::new();
+///
+/// // Run the local task grou[].
+/// local_group.block_on(&mut rt, async move {
+///     let unsend_data = unsend_data.clone();
+///     local::spawn(async move {
+///         println!("{}", unsend_data);
+///         // ...
+///     }).await.unwrap();
+/// });
+/// ```
+pub fn spawn<F>(future: F) -> JoinHandle<F::Output>
+where
+    F: Future + 'static,
+    F::Output: 'static,
+{
+    CURRENT_TASK_SET.with(|current| {
+        let current = current
+            .get()
+            .expect("`local::spawn` called from outside of a local::TaskGroup!");
+        unsafe {
+            let (task, handle) = task::joinable_unsend(future);
+            current.as_ref().schedule(task);
+            handle
+        }
+    })
+}
+
+/// Max number of tasks to poll per tick.
+const MAX_TASKS_PER_TICK: usize = 61;
```

Comment:
This seems pretty magical

Reply:
The current thread runtime and threadpool worker have the same constant...

Reply:
Should we share the constant among all of them? Define it in `src/lib.rs` or something? Also with some docs about why it's set to 61. Probably something for a different PR though. /cc @carllerche

Reply:
i think it was stolen from go, or something